### PR TITLE
Graphql 0.7.0 -> 0.7.2 -> 0.8.0

### PR DESCRIPTION
**Pull request by intellix**
Please fill in this template.

- [x] Prefer to make your PR against the `types-2.0` branch.
- [x] Test the change in your own code.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
- [x] Run `npm run lint -- package-name` if a `tslint.json` is present.

If changing an existing definition:
- [x] Provide a URL to  documentation or source code which provides context for the suggested changes: <https://github.com/graphql/graphql-js/compare/v0.7.0...v0.7.2>
- [x] Increase the version number in the header if appropriate.

### CODE REVIEW

**Comment by intellix** [comment]
Was looking at updating everything, but the changes are huge, so figured it's better to diff the versions and make the changes as I go along. Will attempt to do the others as well

**Comment by intellix** [comment]
Well, that took longer than I thought... Would love someone to check through that to make sure I haven't cocked up anywhere

**Comment by intellix** [comment]
@TonyPythoneer @calebmer

File: graphql/index.d.ts [hunk 1]
```diff
@@ -2620,14 +2791,14 @@ declare module "graphql/utilities/TypeInfo" {
...
-        ObjectValue
+        ValueNode,
+        VariableNode,
+        ListValueNode,
+        ObjectValueNode
     } from 'graphql/language/ast';
 
     function valueFromAST(
-        valueAST: Value, 
+        valueNode: ValueNode,
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.
https://github.com/graphql/graphql-js/blob/v0.8.0/src/utilities/valueFromAST.js#L53

File: graphql/index.d.ts [hunk 2]
```diff
@@ -2603,13 +2774,13 @@ declare module "graphql/utilities/typeComparators" {
 }
 
 declare module "graphql/utilities/typeFromAST" {
-    import { Type } from 'graphql/language/ast';
+    import { TypeNode } from 'graphql/language/ast';
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.
https://github.com/graphql/graphql-js/blob/v0.8.0/src/utilities/typeFromAST.js#L13

File: graphql/index.d.ts [hunk 3]
```diff
@@ -2603,13 +2774,13 @@ declare module "graphql/utilities/typeComparators" {
...
     import { GraphQLType, GraphQLNullableType } from 'graphql/type/definition';
     import { GraphQLSchema } from 'graphql/type/schema';
 
     function typeFromAST(
         schema: GraphQLSchema,
-        inputTypeAST: Type
+        typeNode: TypeNode
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK
https://github.com/graphql/graphql-js/blob/v0.8.0/src/utilities/typeFromAST.js#L21

File: graphql/index.d.ts [hunk 4]
```diff
@@ -2538,27 +2706,30 @@ declare module "graphql/utilities/isValidLiteralValue" {
...
+    function printType(type: GraphQLType): string
 }
 
 declare module "graphql/utilities/separateOperations" {
     import {
-        Document,
-        OperationDefinition,
+        DocumentNode,
+        OperationDefinitionNode,
     } from 'graphql/language/ast';
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.
https://github.com/graphql/graphql-js/blob/v0.8.0/src/utilities/separateOperations.js#L15

File: graphql/index.d.ts [hunk 5]
```diff
@@ -2538,27 +2706,30 @@ declare module "graphql/utilities/isValidLiteralValue" {
...
 
     function separateOperations(
-        documentAST: Document
-    ): { [operationName: string]: Document }
+        documentAST: DocumentNode
+    ): { [operationName: string]: DocumentNode }
 }
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.
https://github.com/graphql/graphql-js/blob/v0.8.0/src/utilities/separateOperations.js#L23-L25

File: graphql/index.d.ts [hunk 6]
```diff
@@ -2538,27 +2706,30 @@ declare module "graphql/utilities/isValidLiteralValue" {
...
     function isValidLiteralValue(
         type: GraphQLInputType,
-        valueAST: Value
+        valueNode: ValueNode
     ): Array<string>
 }
 
 declare module "graphql/utilities/schemaPrinter" {
     import { GraphQLSchema } from 'graphql/type/schema';
+    import { GraphQLType } from 'graphql/type/definition';
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.
https://github.com/graphql/graphql-js/blob/v0.8.0/src/utilities/schemaPrinter.js#L17

File: graphql/index.d.ts [hunk 7]
```diff
@@ -2538,27 +2706,30 @@ declare module "graphql/utilities/isValidLiteralValue" {
...
 
     function printSchema(schema: GraphQLSchema): string;
 
     function printIntrospectionSchema(schema: GraphQLSchema): string;
+
+    function printType(type: GraphQLType): string
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.
https://github.com/graphql/graphql-js/blob/v0.8.0/src/utilities/schemaPrinter.js#L139

File: graphql/index.d.ts [hunk 8]
```diff
@@ -2526,7 +2694,7 @@ declare module "graphql/utilities/isValidJSValue" {
 }
 
 declare module "graphql/utilities/isValidLiteralValue" {
-    import { Value } from 'graphql/language/ast';
+    import { ValueNode } from 'graphql/language/ast';
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.
https://github.com/graphql/graphql-js/blob/v0.8.0/src/utilities/isValidLiteralValue.js#L13

File: graphql/index.d.ts [hunk 9]
```diff
@@ -2538,27 +2706,30 @@ declare module "graphql/utilities/isValidLiteralValue" {
      */
     function isValidLiteralValue(
         type: GraphQLInputType,
-        valueAST: Value
+        valueNode: ValueNode
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.
https://github.com/graphql/graphql-js/blob/v0.8.0/src/utilities/isValidLiteralValue.js#L45

File: graphql/index.d.ts [hunk 10]
```diff
@@ -2280,22 +2354,116 @@ declare module "graphql/utilities/extendSchema" {
...
+     */
+    export function findValuesRemovedFromEnums(
+        oldSchema: GraphQLSchema,
+        newSchema: GraphQLSchema
+    ): Array<BreakingChange>;
+}
+
 declare module "graphql/utilities/getOperationAST" {
-    import { Document, OperationDefinition } from 'graphql/language/ast';
+    import { DocumentNode, OperationDefinitionNode } from 'graphql/language/ast';
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.
https://github.com/graphql/graphql-js/blob/v0.8.0/src/utilities/getOperationAST.js#L12

File: graphql/index.d.ts [hunk 11]
```diff
@@ -2280,22 +2354,116 @@ declare module "graphql/utilities/extendSchema" {
...
      * name. If a name is not provided, an operation is only returned if only one is
      * provided in the document.
      */
     export function getOperationAST(
-        documentAST: Document,
+        documentAST: DocumentNode,
         operationName: string
-    ): OperationDefinition;
+    ): OperationDefinitionNode;
 }
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.
https://github.com/graphql/graphql-js/blob/v0.8.0/src/utilities/getOperationAST.js#L20

File: graphql/index.d.ts [hunk 12]
```diff
@@ -2280,22 +2354,116 @@ declare module "graphql/utilities/extendSchema" {
...
+    /**
+     * Given two schemas, returns an Array containing descriptions of any breaking
+     * changes in the newSchema related to removing an entire type.
+     */
+    export function findRemovedTypes(
+        oldSchema: GraphQLSchema,
+        newSchema: GraphQLSchema
+    ): Array<BreakingChange>
+
+    function typeKindName(type: GraphQLNamedType): string
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: Please delete it. It only writes the property has been exported in module.

File: graphql/index.d.ts [hunk 13]
```diff
@@ -2280,22 +2354,116 @@ declare module "graphql/utilities/extendSchema" {
...
+    /**
+     * Given two schemas, returns an Array containing descriptions of any breaking
+     * changes in the newSchema related to changing the type of a type.
+     */
+    export function findTypesThatChangedKind(
+        oldSchema: GraphQLSchema,
+        newSchema: GraphQLSchema
+    ): Array<BreakingChange>
+
+    function typeKindName(type: GraphQLNamedType): string;
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: Please delete it. It only writes the property has been exported in module.

File: graphql/index.d.ts [hunk 14]
```diff
@@ -2253,17 +2327,17 @@ declare module "graphql/utilities/buildClientSchema" {
...
      * concatenate the ASTs together into batched AST, useful for validating many
      * GraphQL source files which together represent one conceptual application.
      */
-    function concatAST(asts: Array<Document>): Document;
+    function concatAST(asts: Array<DocumentNode>): DocumentNode;
 }
 
 declare module "graphql/utilities/extendSchema" {
-    import { Document } from 'graphql/language/ast';
+    import { DocumentNode } from 'graphql/language/ast';
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.

File: graphql/index.d.ts [hunk 15]
```diff
@@ -2280,22 +2354,116 @@ declare module "graphql/utilities/extendSchema" {
      */
     function extendSchema(
         schema: GraphQLSchema,
-        documentAST: Document
+        documentAST: DocumentNode
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.

File: graphql/index.d.ts [hunk 16]
```diff
@@ -2253,17 +2327,17 @@ declare module "graphql/utilities/buildClientSchema" {
...
 declare module "graphql/utilities/concatAST" {
-    import { Document } from 'graphql/language/ast';
+    import { DocumentNode } from 'graphql/language/ast';
     /**
      * Provided a collection of ASTs, presumably each from different files,
      * concatenate the ASTs together into batched AST, useful for validating many
      * GraphQL source files which together represent one conceptual application.
      */
-    function concatAST(asts: Array<Document>): Document;
+    function concatAST(asts: Array<DocumentNode>): DocumentNode;
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.

File: graphql/index.d.ts [hunk 17]
```diff
@@ -2100,6 +2151,25 @@ declare module "graphql/utilities" {
 declare module "graphql/utilities/index" {
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: `module "graphql/utilities/index"` is OK

File: graphql/index.d.ts [hunk 18]
```diff
@@ -1988,7 +2038,8 @@ declare module "graphql/error/formatError" {
 
     type GraphQLFormattedError = {
         message: string,
-        locations: Array<GraphQLErrorLocation>
+        locations: Array<GraphQLErrorLocation>,
+        path: Array<string | number>
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.

File: graphql/index.d.ts [hunk 19]
```diff
@@ -1957,8 +2007,8 @@ declare module "graphql/execution/values" {
      * definitions and list of argument AST nodes.
      */
     function getArgumentValues(
-        argDefs: Array<GraphQLArgument>,
-        argASTs: Array<Argument>,
+        def: GraphQLField | GraphQLDirective,
+        node: FieldNode | DirectiveNode,
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.

File: graphql/index.d.ts [hunk 20]
```diff
@@ -1927,28 +1958,47 @@ declare module "graphql/execution/execute" {
...
+    import { ArgumentNode, FieldNode, DirectiveNode, VariableDefinitionNode } from 'graphql/language/ast';
     /**
      * Prepares an object map of variableValues of the correct type based on the
      * provided variable definitions and arbitrary input. If the input cannot be
      * parsed to match the variable definitions, a GraphQLError will be thrown.
      */
     function getVariableValues(
         schema: GraphQLSchema,
-        definitionASTs: Array<VariableDefinition>,
+        varDefNodes: Array<VariableDefinitionNode>,
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.

File: graphql/index.d.ts [hunk 21]
```diff
@@ -1927,28 +1958,47 @@ declare module "graphql/execution/execute" {
...
+    export const defaultFieldResolver: GraphQLFieldResolver<any>;
 }
 
 declare module "graphql/execution/values" {
-    import { GraphQLInputType, GraphQLArgument } from 'graphql/type/definition';
+    import { GraphQLInputType, GraphQLField, GraphQLArgument } from 'graphql/type/definition';
+    import { GraphQLDirective } from 'graphql/type/directives';
     import { GraphQLSchema } from 'graphql/type/schema';
-    import { Argument, VariableDefinition } from 'graphql/language/ast';
+    import { ArgumentNode, FieldNode, DirectiveNode, VariableDefinitionNode } from 'graphql/language/ast';
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: `ArgumentNode` is unused, it can delete.

File: graphql/index.d.ts [hunk 22]
```diff
@@ -1912,8 +1943,8 @@ declare module "graphql/execution/execute" {
      * query, `errors` is null if no errors occurred, and is a
      * non-empty array if an error occurred.
      */
-    interface ExecutionResult {
-        data: Object;
+    export interface ExecutionResult {
+        data: {[key: string]: any};
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.
https://github.com/graphql/graphql-js/blob/v0.8.0/src/execution/execute.js#L102

File: graphql/index.d.ts [hunk 23]
```diff
@@ -1900,9 +1931,9 @@ declare module "graphql/execution/execute" {
      */
     interface ExecutionContext {
         schema: GraphQLSchema;
-        fragments: { [key: string]: FragmentDefinition };
+        fragments: { [key: string]: FragmentDefinitionNode };
         rootValue: any;
-        operation: OperationDefinition;
+        operation: OperationDefinitionNode;
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.
https://github.com/graphql/graphql-js/blob/v0.8.0/src/execution/execute.js#L86

File: graphql/index.d.ts [hunk 24]
```diff
@@ -1877,20 +1907,21 @@ declare module "graphql/execution" {
 }
 
 declare module "graphql/execution/index" {
-    export { execute } from 'graphql/execution/execute';
+    export { execute, defaultFieldResolver, responsePathAsArray, ExecutionResult } from 'graphql/execution/execute';
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.

File: graphql/index.d.ts [hunk 25]
```diff
@@ -1765,19 +1795,19 @@ declare module "graphql/validation/specifiedRules" {
 declare module "graphql/validation/validate" {
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: `module "graphql/validation/validate"` is OK.

File: graphql/index.d.ts [hunk 26]
```diff
@@ -1647,12 +1677,12 @@ declare module "graphql/type/introspection" {
...
-     * Note that these are GraphQLFieldDefinition and not GraphQLFieldConfig,
+     * Note that these are GraphQLField and not GraphQLFieldConfig,
      * so the format for args is different.
      */
-    const SchemaMetaFieldDef: GraphQLFieldDefinition;
-    const TypeMetaFieldDef: GraphQLFieldDefinition;
-    const TypeNameMetaFieldDef: GraphQLFieldDefinition;
+    const SchemaMetaFieldDef: GraphQLField;
+    const TypeMetaFieldDef: GraphQLField;
+    const TypeNameMetaFieldDef: GraphQLField;
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.

File: graphql/index.d.ts [hunk 27]
```diff
@@ -1471,15 +1501,15 @@ declare module "graphql/type/definition" {
...
         name: string;
         type: GraphQLInputType;
         defaultValue?: any;
         description?: string;
     }
 
-    export interface GraphQLInputFieldDefinitionMap {
-        [fieldName: string]: GraphQLInputFieldDefinition;
+    export interface GraphQLInputFieldMap {
+        [fieldName: string]: GraphQLInputField;
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.

File: graphql/index.d.ts [hunk 28]
```diff
@@ -980,14 +1007,15 @@ declare module "graphql/type/index" {
         TypeNameMetaFieldDef,
     } from 'graphql/type/introspection';
 
+    export { DirectiveLocationEnum } from 'graphql/type/directives';
 }
 
 declare module "graphql/type/definition" {
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: `"graphql/type/definition"` is ok

File: graphql/index.d.ts [hunk 29]
```diff
@@ -980,14 +1007,15 @@ declare module "graphql/type/index" {
         TypeNameMetaFieldDef,
     } from 'graphql/type/introspection';
 
+    export { DirectiveLocationEnum } from 'graphql/type/directives';
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.
https://github.com/graphql/graphql-js/blob/v0.8.0/src/type/index.js#L84

File: graphql/index.d.ts [hunk 30]
```diff
@@ -889,6 +915,7 @@ declare module "graphql/language/visitor" {
         FloatValue: number[];
         StringValue: string[];
         BooleanValue: boolean[];
+        NullValue: null[],
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.
https://github.com/graphql/graphql-js/blob/v0.8.0/src/language/visitor.js#L30

File: graphql/index.d.ts [hunk 31]
```diff
@@ -801,7 +819,7 @@ declare module "graphql/language/location" {
 }
 
 declare module "graphql/language/parser" {
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: `module "graphql/language/parser"` is OK.

File: graphql/index.d.ts [hunk 32]
```diff
@@ -664,6 +681,7 @@ declare module "graphql/language/kinds" {
     const FLOAT: 'FloatValue';
     const STRING: 'StringValue';
     const BOOLEAN: 'BooleanValue';
+    const NULL: 'NullValue';
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.
https://github.com/graphql/graphql-js/blob/v0.8.0/src/language/kinds.js#L37

File: graphql/index.d.ts [hunk 33]
```diff
@@ -102,12 +110,32 @@ declare module "graphql" {
...
+        IntrospectionQuery,
+        IntrospectionScalarType,
+        IntrospectionSchema,
+        IntrospectionType,
+        IntrospectionTypeRef,
+        IntrospectionUnionType,
     } from 'graphql/utilities';
 }
 
 declare module "graphql/graphql" {
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: This module is ok.

**Comment by TonyPythoneer** [comment]
Hi, @intellix

Thank you for your contribution. I appreciate you bring these changes. It makes this type definition up to date. I welcome you to join this. Let's make this better.

File: graphql/index.d.ts (hunk 13 above)
Comment (TonyPythoneer) [inline graphql/index.d.ts]: @intellix here needs to fix.

File: graphql/index.d.ts (hunk 21 above)
Comment (TonyPythoneer) [inline graphql/index.d.ts]: @intellix here needs to fix.

**Comment by intellix** [comment]
Fixed those 3 lines, thanks for the thorough checks :)

**Comment by TonyPythoneer** [comment]
Ok, it can be merged.

**Comment by DxCx** [comment]
Kudos!
Thanks alot im sure lots of time was spent in here..

**Comment by TonyPythoneer** [comment]
@DxCx Yes, it's @intellix's merits. He makes this update latest! :tada:

I took 2 hours to check this. The main change reason is that source code changes naming style of the variables/functions. I believe when @intellix did the comparison between 0.7.0 and 0.8.0, it's a huge hard task.

Would you join here? Let's make this better!  :smile:

**Comment by intellix** [comment]
Should do 0.8.2 after this is merged. Don't wanna hold back this PR though :) 
Diff: https://github.com/graphql/graphql-js/compare/v0.8.0...v0.8.2
One thing I wasn't sure on are the * in Flow. I'm not seeing any documentation about them anywhere:

```js
export const defaultFieldResolver: GraphQLFieldResolver<any, *>;
```

**Comment by DxCx** [comment]
Yeah @TonyPythoneer after it is merged and released im planing to change apollostack repos to use it instead of typed-graphql
And if more symbols will be missing ill post a PR to align them.. :)

**Comment by TonyPythoneer** [comment]
@intellix You can check out and create a new branch based on this one if you are available to do v0.8.2. I think it's okay. :smile: 

I don't know what '*' is in Flow, too. I have issued a question on `graphql`. I hope we can get answer soon. :smile: 

When I made graphql type definition, I had met the same issue. Flow has used many special type syntax, such as `mixed`, `$key`, etc. It's not easy to understand directly, not like typescript which it's simply and clearly. So, it make me hate Flow more. :sob:

**Comment by TonyPythoneer** [comment]
I think this type definition still needs to improve, for example, to meet types-2.0 style and add test cases.
I am planning to do those after 0.8.2 gets merged. :smile:

**Pull request merged**